File: brains/bayes.py
from __future__ import division
import random
from brains import ProbabilityMap

class Bayes(object):
    "Naive bayes classifier."
    def __init__(s):
        s.pmap = ProbabilityMap()

    def train(s, klass, features):
        "Adds the sample with given class and features."
        s.pmap.update(klass, features)

    @property
    def count(s):
        "Returns the number of samples trained by this classifier."
        return s.pmap.count

    def classify(s, features):
        "Classifies the sample with features. Returns most probable class"
        ranked = []
        for klass in s.pmap.pc.iterkeys():
            p = 1
            for feat, val in features.items():
                c = s.pmap.getpfc(feat, val, klass)
                if c:
                    p *= c / s.pmap.count
                else:
                    p *= 0.01 / s.pmap.count # TODO: adjust this
            ranked.append((p * s.pmap.pc.get(klass, 0.0), klass))
        ranked.sort()
        return ranked[-1][1]

File: lib/brains/bayes.rb
module Brains

# Naive bayes classifier.
class Bayes

  attr_reader :count, :pc, :pf

  def initialize
    @pmap = ProbabilityMap.new
  end

  def train(points)
    @pmap.train(points)
  end

  def classify(features)
    probs = {}
    @pmap.pc.keys.each do |klass|
      p = 1.0
      features.each do |f,v|
        p *= @pmap.pf.fetch(f, {}).fetch(v, {}).fetch(klass, 0.01) / @pmap.count
      end
      probs[klass] = p
    end
    probs.to_a.sort { |a,b| b.last <=> a.last }.first.first
  end

end

end
